src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs

```diff
@@ -3086,35 +3113,59 @@ mergeEqualPs = go (empty, [])
     go (_, [])       [] = return []
     go (_, []) (p : ps) = (p :) <$> mergeEqualPs ps
 
--- does not check pattern linearity
+
+-- | Scope-check a 'C.LHSCore' (of possibly a 'C.DisplayForm').
+
+data CLHSCore = CLHSCore
+  Bool
+    -- ^ Are we checking the left hand side of a 'C.DisplayForm'?
+  C.LHSCore
+    -- ^ The lhs to scope-check.
+
+-- | Scope-check a 'C.LHSCore' not of a 'C.DisplayForm'.
+
 instance ToAbstract C.LHSCore where
-    type AbsOfCon C.LHSCore = (A.LHSCore' C.Expr)
+  type AbsOfCon C.LHSCore = A.LHSCore' C.Expr
+
+  toAbstract = toAbstract . CLHSCore False
+
+-- does not check pattern linearity
+instance ToAbstract CLHSCore where
+  type AbsOfCon CLHSCore = A.LHSCore' C.Expr
+
+  toAbstract (CLHSCore displayLhs core0) = case core0 of
 
-    toAbstract (C.LHSHead x ps) = do
-        x <- withLocalVars $ do
+    C.LHSHead x ps -> do
+        x <- withLocalVars do
           setLocalVars []
           toAbstract (OldName x)
-        A.LHSHead x <$> do mergeEqualPs =<< toAbstract ps
-    toAbstract (C.LHSProj d ps1 l ps2) = do
+        ps <- toAbstract $ (fmap . fmap . fmap) (CPattern displayLhs) ps
+        A.LHSHead x <$> mergeEqualPs ps
+
+    C.LHSProj d ps1 core ps2 -> do
         unless (null ps1) $ typeError $ IllformedProjectionPatternConcrete (foldl C.AppP (C.IdentP True d) ps1)
         ds <- resolveName d >>= \case
           FieldName ds -> return $ fmap anameName ds
           UnknownName  -> notInScopeError d
           _            -> typeError $ CopatternHeadNotProjection d
-        A.LHSProj (AmbQ ds) <$> toAbstract l <*> (mergeEqualPs =<< toAbstract ps2)
-    toAbstract (C.LHSWith core wps ps) = do
-      liftA2 A.lhsCoreApp
-        (liftA2 A.lhsCoreWith
-          (toAbstract core)
-          (fmap defaultArg <$> toAbstract wps))
-        (toAbstract ps)
+        core <- toAbstract $ (fmap . fmap) (CLHSCore displayLhs) core
+        ps2  <- toAbstract $ (fmap . fmap . fmap) (CPattern displayLhs) ps2
+        A.LHSProj (AmbQ ds) core <$> mergeEqualPs ps2
+
+    C.LHSWith core wps ps -> do  -- N.B.: This is an ApplicativeDo.
```

Comment:
Why is it important that this is an ApplicativeDo? I would be a little concerned if the semantics relies on this to be the case!

Reply:
I can take this comment away, if it is irritating.  
This was a nested `liftA2`, but rather unreadable.  
`ApplicativeDo` shouldn't change the semantics, don't even know whether it makes things more performant.

Reply:
I would get rid of the comment and the options pragma at the top. We don't care about restricting to the applicative interface anywhere else, nor do we have any evidence that it would be beneficial.

Reply:
Done.